e2e/nomostest/sync.go

```diff
@@ -263,3 +266,58 @@ func statusHasSyncDirAndNoErrors(status v1beta1.Status, sourceType configsync.So
 	}
 	return nil
 }
+
+func resourceGroupHasReconciled(sourceHash string) testpredicates.Predicate {
+	return func(o client.Object) error {
+		if o == nil {
+			return testpredicates.ErrObjectNotFound
+		}
+		rg, ok := o.(*v1alpha1.ResourceGroup)
+		if !ok {
+			return testpredicates.WrongTypeErr(o, &v1alpha1.ResourceGroup{})
+		}
+
+		// Status is disabled for this ResourceGroup, skip all status checks
+		if root.IsStatusDisabled(rg) {
+			return nil
+		}
+
+		if len(rg.Spec.Resources) != len(rg.Status.ResourceStatuses) {
+			return fmt.Errorf("length of spec.resources (%d) does not equal length of status.resourceStatuses (%d)",
+				len(rg.Spec.Resources), len(rg.Status.ResourceStatuses))
+		}
+
+		resourceCount := make(map[v1alpha1.ObjMetadata]int)
+		for _, s := range rg.Status.ResourceStatuses {
+			if err := resourceStatusIsCurrent(s, sourceHash); err != nil {
+				return fmt.Errorf("object %s is not current: %w", s.ObjMetadata, err)
+			}
+			resourceCount[s.ObjMetadata]++
+		}
+
+		for _, o := range rg.Spec.Resources {
+			if count, ok := resourceCount[o]; ok && count > 0 {
+				resourceCount[o]--
+			} else {
+				return fmt.Errorf("spec.resources does not equal status.resourceStatuses")
+			}
+		}
+		return nil
+	}
+}
+
+func resourceStatusIsCurrent(rs v1alpha1.ResourceStatus, sourceHash string) error {
+	if rs.Status != v1alpha1.Current {
+		return fmt.Errorf("resourceStatus.status is not %s. Got %s", v1alpha1.Current, rs.Status)
+	}
+	if rs.Reconcile != v1alpha1.ReconcileSucceeded {
+		return fmt.Errorf("resourceStatus.reconcile is not %s. Got %s", v1alpha1.ReconcileSucceeded, rs.Reconcile)
+	}
+	if rs.Actuation != v1alpha1.ActuationSucceeded {
+		return fmt.Errorf("resourceStatus.actuation is not %s. Got %s", v1alpha1.ActuationSucceeded, rs.Actuation)
+	}
+	if rs.SourceHash != status.TruncateSourceHash(sourceHash) {
```

Comment:
Huh. I didn't even know these had a source hash field. Nifty.

So I guess you can tell if they're all on the same commit pretty easy, you just can't tell what commit it's supposed to be from the spec.

I guess if we have this in the status, maybe we should add a SourceHash to the spec, instead of adding a source-commit annotation. That would make it possible to check that the status matches the spec without external knowledge.

Guess we can do that in a different PR, but we're gonna need approval from @janetkuo